### PR TITLE
adding smoke tests

### PR DESCRIPTION
**Pull request by Prateeknandle**
Signed-off-by: Prateeknandle <prateeknandle@gmail.com>

### CODE REVIEW

File: tests/smoke/smoke_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/accuknox/auto-policy-discovery/src/types"
+	"github.com/kubearmor/KubeArmor/tests/util"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/rs/zerolog/log"
+)
+
+var _ = BeforeSuite(func() {
+	//install wordpress-mysql app
+	err := util.K8sApply([]string{"res/wordpress-mysql-deployment.yaml"})
```
Comment (seswarrajan) [inline tests/smoke/smoke_test.go]: My suggestion is to Deploy DE and then deploy wordpress-mysql app for test.

**Review by seswarrajan — CHANGES_REQUESTED** [review]
1. Kindly deploy DE then deploy WP-MySQL demo app
2. Why is the go.mod file looks new. There should be an update on old one and not new one to be added. Is it that go.mod not added previously ?

File: tests/smoke/smoke_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		log.Error().Msgf("Failed to apply the `karmor discover` command : %v", err)
+	}
+	err = json.Unmarshal(cmd, &policy)
+	if err != nil {
+		log.Error().Msgf("Failed to unmarshal the policy : %v", err)
+	}
+	return policy, err
+}
+
+var _ = Describe("Smoke", func() {
```
Comment (rksharma95) [inline tests/smoke/smoke_test.go:100]: before running a test, make sure that `discovery-engine` and `wordpress-app` is up and running using `BeforeEach()`

File: tests/smoke/smoke_test.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	Describe("Auto Policy Discovery", func() {
+		It("test", func() {
+			policy, err := discover("wordpress-mysql", "app=wordpress")
+			Expect(err).To(BeNil())
+			Expect(policy.APIVersion).To(Equal("security.kubearmor.com/v1"))
+			Expect(policy.Kind).To(Equal("KubeArmorPolicy"))
+			Expect(policy.Metadata["namespace"]).To(Equal("wordpress-mysql"))
+			Expect(policy.Spec.Action).To(Equal("Allow"))
+			Expect(policy.Spec.Selector.MatchLabels["app"]).To(Equal("wordpress"))
+			Expect(policy.Spec.File.MatchDirectories[0].Dir).To(Equal("/usr/lib/x86_64-linux-gnu/"))
```
Comment (rksharma95) [inline tests/smoke/smoke_test.go]: expecting data to be at fixed position i.e. at index 0 in this case, may lead to an issue and key to flakyness.

File: tests/smoke/smoke_test.go (hunk 1 above)
Comment (Prateeknandle) [reply tests/smoke/smoke_test.go]: go.mod was not present earlier

File: tests/smoke/smoke_test.go (hunk 3 above)
Comment (Prateeknandle) [reply tests/smoke/smoke_test.go]: @rksharma95 what could be a possible solution for this, storing values in map with assigning a specific key, as MatchLabels is a map?

Comment (Prateeknandle) [reply tests/smoke/smoke_test.go]: also does these values change on different host?

Comment (rksharma95) [reply tests/smoke/smoke_test.go]: > also can these values be changed on different host?

i don't think so, to be more confident we can target wordpress app specific policies they will not change.

> @rksharma95 what could be a possible solution for this, storing values in map with assigning a specific key, as MatchLabels is a map?

we can traverse slices i.e. MatchDirectories[ ], MatchPaths[ ] and look for the target values.

Comment (Prateeknandle) [reply tests/smoke/smoke_test.go:64]: > we can traverse slices i.e. MatchDirectories[ ], MatchPaths[ ] and look for the target values.

@rksharma95 can we do something like this
```
for i := range policy.Spec.File.MatchDirectories {
				if policy.Spec.File.MatchDirectories[i].Dir == "/etc/" {
					value = "/etc/"
				}
			}
			Expect(value).To(Equal("/etc/"))
```

File: tests/smoke/smoke_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		getWpsqlPod("discovery-engine-",
+			"container.apparmor.security.beta.kubernetes.io/discovery-engine: localhost/kubearmor-accuknox-agents-discovery-engine-discovery-engine", "accuknox-agents")
+	})
+
+	AfterEach(func() {
+		util.KarmorLogStop()
+		err := util.DeleteAllKsp()
+		Expect(err).To(BeNil())
+		// wait for policy deletion
+		time.Sleep(5 * time.Second)
```
Comment (rksharma95) [inline tests/smoke/smoke_test.go:116]: it's not needed here, we are not applying any security policy, 
is there any other purpose for the sleep here?

**Comment by rksharma95** [comment]
test is failing when i tried it locally, @Prateeknandle can you verify if it's working?
`{"level":"error","time":"2023-01-11T10:23:56+05:30","message":"Failed to unmarshal the policy : invalid character '{' after top-level value"}
`

File: tests/smoke/smoke_test.go (hunk 4 above)
Comment (Prateeknandle) [reply tests/smoke/smoke_test.go:116]: I thought if someone had a kubearmor policy already in the cluster then it'll be removed. Should I remove this?

**Comment by Prateeknandle** [comment]
> test is failing when i tried it locally, @Prateeknandle can you verify if it's working? `{"level":"error","time":"2023-01-11T10:23:56+05:30","message":"Failed to unmarshal the policy : invalid character '{' after top-level value"} `

@rksharma95 I'm not getting any error like this, I hope you're using `karmor discover -n wordpress-mysql -l "app=wordpress" -f json`, this command with json output not with yaml.

If you're using the same code then I don't understand what's the problem because I'm getting the expected result

**Comment by rksharma95** [comment]
> > test is failing when i tried it locally, @Prateeknandle can you verify if it's working? `{"level":"error","time":"2023-01-11T10:23:56+05:30","message":"Failed to unmarshal the policy : invalid character '{' after top-level value"} `
> 
> @rksharma95 I'm not getting any error like this, I hope you're using `karmor discover -n wordpress-mysql -l "app=wordpress" -f json`, this command with json output not with yaml.
> 
> If you're using the same code then I don't understand what's the problem because I'm getting the expected result

i think we should have CI workflow for testing first to handle this issue better.

**Review by seswarrajan — APPROVED** [review]
LGTM

**Comment by Prateeknandle** [comment]
will open the new pr, blundered this during rebase

**Comment by Prateeknandle** [comment]
will open new pr, blundered this one during rebase

**Pull request closed without merging**